File: native/native.tests/testData/framework/objcexport/interfaceMethodNameMangling.kt
@file:OptIn(kotlin.experimental.ExperimentalObjCName::class)

package objcNameMangling

interface InterfaceNameManglingI1 {
    val clashingProperty: Int

    fun clashingMethod(): Int

    @ObjCName(name = "interfaceClashingMethodWithObjCNameInI1")
    fun clashingMethodWithObjCNameInI1(): Int

    fun interfaceClashingMethodWithObjCNameInI2(): Int

    @ObjCName(name = "interfaceClashingMethodWithObjCNameInBoth")
    fun clashingMethodWithObjCNameInBoth(): Int
}

fun i1() = object : InterfaceNameManglingI1 {
    override val clashingProperty: Int
        get() = 1

    override fun clashingMethod(): Int = 2

    override fun clashingMethodWithObjCNameInI1(): Int = 3

    override fun interfaceClashingMethodWithObjCNameInI2(): Int = 4

    override fun clashingMethodWithObjCNameInBoth(): Int = 5
}

interface InterfaceNameManglingI2 {
    val clashingProperty: Any

    fun clashingMethod(): Any

    fun interfaceClashingMethodWithObjCNameInI1(): Any

    @ObjCName(name = "interfaceClashingMethodWithObjCNameInI2")
    fun clashingMethodWithObjCNameInI2(): Any

    @ObjCName(name = "interfaceClashingMethodWithObjCNameInBoth")
    fun clashingMethodWithObjCNameInBoth(): Any
}

fun i2() = object : InterfaceNameManglingI2 {
    override val clashingProperty: Any
        get() = "one"

    override fun clashingMethod(): Any = "two"

    override fun interfaceClashingMethodWithObjCNameInI1(): Any = "three"

    override fun clashingMethodWithObjCNameInI2(): Any = "four"

    override fun clashingMethodWithObjCNameInBoth(): Any = "five"
}

class InterfaceNameManglingC1 {
    val clashingProperty: String = "one"

    fun clashingMethod(): String = "two"
}

final class InterfaceNameManglingC2 {
    val clashingProperty: Int = 1

    fun clashingMethod(): Int = 2
}

fun o1() = InterfaceNameManglingC1()

fun o2() = InterfaceNameManglingC2()

File: native/native.tests/testData/framework/objcexport/interfaceMethodNameMangling.swift
import Kt

private func test1() throws {
    let i1 = InterfaceMethodNameManglingKt.i1()
    let i2 = InterfaceMethodNameManglingKt.i2()
    let o1 = InterfaceMethodNameManglingKt.o1()
    let o2 = InterfaceMethodNameManglingKt.o2()

#if DISABLE_MEMBER_NAME_MANGLING || DISABLE_INTERFACE_METHOD_NAME_MANGLING
    try assertEquals(actual: i1.clashingProperty, expected: 1)
    try assertEquals(actual: i1.clashingMethod(), expected: 2)
    try assertEquals(actual: i1.interfaceClashingMethodWithObjCNameInI1(), expected: 3)
    try assertEquals(actual: i1.interfaceClashingMethodWithObjCNameInI2(), expected: 4)
    try assertEquals(actual: i1.interfaceClashingMethodWithObjCNameInBoth(), expected: 5)

    try assertEquals(actual: i2.clashingProperty as! String, expected: "one")
    try assertEquals(actual: i2.clashingMethod() as! String, expected: "two")
    try assertEquals(actual: i2.interfaceClashingMethodWithObjCNameInI1() as! String, expected: "three")
    try assertEquals(actual: i2.interfaceClashingMethodWithObjCNameInI2() as! String, expected: "four")
    try assertEquals(actual: i2.interfaceClashingMethodWithObjCNameInBoth() as! String, expected: "five")

    try assertEquals(actual: o1.clashingProperty, expected: "one")
    try assertEquals(actual: o1.clashingMethod(), expected: "two")
#endif
    try assertEquals(actual: o2.clashingProperty, expected: 1)
    try assertEquals(actual: o2.clashingMethod(), expected: 2)
}

class InterfaceMethodNameManglingTests : SimpleTestProvider {
    override init() {
        super.init()

        test("Test1", test1)
    }
}